src/Lachain.Consensus/AbstractProtocol.cs

```diff
@@ -144,8 +147,13 @@ public void ReceiveMessage(MessageEnvelope message)
         {
             lock (_queueLock)
             {
-                if (Terminated){}
-            //        Logger.LogTrace($"{Id}: got message after termination"); 
+                if (Terminated)
+                {
+                    // we should return here instead of enqueueing messages
+                    // because once terminated, the messages are not being processed anymore
+                    // so the queue will just get large unnecessarily
+                    return;
```

Comment:
The same: can we get some deadlock here due to the absence of the Monitor.Pulse?

Reply:
Here adding Monitor.Pulse(_queueLock); before return will work, right?

Reply:
I don't sure. It can unlock the thread which tries to read from queue and gets new exception

Reply:
The queue is read only in Start() method, and it is private so not accessible from outside, also at this point the incoming message is not enqueued yet, so I think it should not be a problem.

Reply:
Yes,  you are right